### PR TITLE
[CIT-459] Use the batcher to batch collectd message

### PR DESCRIPTION
**Pull request by didier-wenzek**
## Proposed changes

* Use the crate `common/batcher` to batch the collectd messages.
* This first draft is focused on getting things work.
* One still needs to improve configuration, code clarity.

## Types of changes

What types of changes does your code introduce to `thin-edge.io`?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Improvement (general improvements like code refactorings that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
<br/>

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ ] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [ ] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: mapper/tedge_mapper/src/collectd_mapper/batcher.rs [hunk 1]
```diff
@@ -1,27 +1,46 @@
...
 }
 
 impl MessageBatch {
+    pub fn thin_edge_json_bytes(
+        messages: Vec<CollectdMessage>,
+    ) -> Result<Payload, DeviceMonitorError> {
+        let mut messages = messages.into_iter();
+
+        if let Some(first_message) = messages.next() {
+            let timestamp = first_message.timestamp.with_timezone(Local::now().offset());
```
[the remark this answers is not in the document]
Comment (didier-wenzek) [reply mapper/tedge_mapper/src/collectd_mapper/batcher.rs:25]: I would have prefer to convert the timestamp into a `Datetime<FixOffset>, *without* getting the current time.

File: mapper/tedge_mapper/src/collectd_mapper/monitor.rs [hunk 2]
```diff
@@ -61,50 +58,85 @@ impl DeviceMonitor {
...
-            TopicFilter::new(self.device_monitor_config.mqtt_source_topic)?.qos(QoS::AtMostOnce),
-            Arc::new(WallClock),
+        let mqtt_client: Arc<dyn MqttClient> = Arc::new(
+            Client::connect(self.device_monitor_config.mqtt_client_id, &mqtt_config).await?,
         );
-        let join_handle1 = tokio::task::spawn(async move {
-            match message_batch_producer.run().await {
+
+        let batch_config = BatchConfigBuilder::new()
+            .event_jitter(50)
```
[the remark this answers is not in the document]
Comment (didier-wenzek) [reply mapper/tedge_mapper/src/collectd_mapper/monitor.rs:71]: Updated so the 3 parameters are read from the config.

File: mapper/tedge_mapper/src/collectd_mapper/monitor.rs [hunk 3]
```diff
@@ -61,50 +58,85 @@ impl DeviceMonitor {
...
+                                error!("Error while processing a collectd message: {}", err);
+                            }
+                        }
+                        Err(err) => {
+                            error!("Error while decoding a collectd message: {}", err);
+                        }
+                    },
+                    None => {
+                        //If the message batching loop returns, it means the MQTT connection has closed
+                        error!("MQTT connection closed. Retrying...");
```
[the remark this answers is not in the document]
Comment (didier-wenzek) [reply mapper/tedge_mapper/src/collectd_mapper/monitor.rs:105]: It happens that `None` is returned *only* when the channel is closed on program exit and never when the MQTT connection is temporary lost.

Hence, one just have to return here.

[the remark this answers is not in the document]
Comment (didier-wenzek) [reply mapper/tedge_mapper/src/collectd_mapper/monitor.rs:105]: No, this is a bit more involved. One needs to propagate to the end of stream to the batch driver thread and the publisher thread.

File: mapper/tedge_mapper/src/collectd_mapper/monitor.rs [hunk 4]
```diff
@@ -61,50 +58,85 @@ impl DeviceMonitor {
...
-            Arc::new(WallClock),
+        let mqtt_client: Arc<dyn MqttClient> = Arc::new(
+            Client::connect(self.device_monitor_config.mqtt_client_id, &mqtt_config).await?,
         );
-        let join_handle1 = tokio::task::spawn(async move {
-            match message_batch_producer.run().await {
+
+        let batch_config = BatchConfigBuilder::new()
+            .event_jitter(50)
+            .delivery_jitter(20)
```
[the remark this answers is not in the document]
Comment (didier-wenzek) [reply mapper/tedge_mapper/src/collectd_mapper/monitor.rs]: Using event jitter of 500ms and delivery jitter of 200ms gives meaningful groups. => Taking these as the default.

```
[tedge/measurements] {"time":"2021-09-02T16:51:21.285000085+01:00","cpu":{"percent-active":1.25},"memory":{"percent-used":1.25214329688669}}
[tedge/measurements] {"time":"2021-09-02T16:51:22.286000013+01:00","cpu":{"percent-active":0.25},"memory":{"percent-used":1.25590837628407}}
[tedge/measurements] {"time":"2021-09-02T16:51:23.286000013+01:00","memory":{"percent-used":1.25590837628407},"cpu":{"percent-active":0.0}}
[tedge/measurements] {"time":"2021-09-02T16:51:24.286000013+01:00","memory":{"percent-used":1.25580661738144},"cpu":{"percent-active":0.0}}
[tedge/measurements] {"time":"2021-09-02T16:51:25.286000013+01:00","memory":{"percent-used":1.25580661738144},"cpu":{"percent-active":0.0},"df-root":{"percent_bytes-used":64.2817306518555}}
[tedge/measurements] {"time":"2021-09-02T16:51:26.286000013+01:00","memory":{"percent-used":1.25580661738144},"cpu":{"percent-active":3.0}}
[tedge/measurements] {"time":"2021-09-02T16:51:27.286000013+01:00","cpu":{"percent-active":0.25},"memory":{"percent-used":1.25580661738144}}
```

**Review by makr11st — APPROVED** [review]
LGTM

File: mapper/tedge_mapper/src/collectd_mapper/monitor.rs [hunk 5]
```diff
@@ -11,8 +11,8 @@ use mqtt_client::{QoS, Topic, TopicFilter};
 const DEFAULT_HOST: &str = "localhost";
 const DEFAULT_PORT: u16 = 1883;
 const DEFAULT_MQTT_CLIENT_ID: &str = "collectd-mapper";
-const DEFAULT_BATCHING_WINDOW: u32 = 200;
-const DEFAULT_MAXIMUM_MESSAGE_DELAY: u32 = 50;
+const DEFAULT_BATCHING_WINDOW: u32 = 500;
+const DEFAULT_MAXIMUM_MESSAGE_DELAY: u32 = 500;
```
Comment (albinsuresh) [inline mapper/tedge_mapper/src/collectd_mapper/monitor.rs:15]: You really wanted a max message delay of `500` or `200`? One of your earlier comments mentioned 200. So just checking.

**Review by albinsuresh — APPROVED** [review]

File: mapper/tedge_mapper/src/collectd_mapper/monitor.rs (hunk 5 above)
Comment (didier-wenzek) [reply mapper/tedge_mapper/src/collectd_mapper/monitor.rs:15]: Oops

**Pull request merged**